### PR TITLE
[Core] Refactoring disaggregated prefilling/decoding using Mooncake Transfer Engine

### PR DESCRIPTION
**Pull request by alogfans**
This PR is related to #10727, as well a continuation of [PR #8498](https://github.com/vllm-project/vllm/pull/8498), which uses [Mooncake's](https://github.com/kvcache-ai/mooncake) Transfer Engine for KVCache transfer instead of NCCL. 

[Mooncake](https://github.com/kvcache-ai/mooncake) is a KVCache-centric disaggregated architecture for LLM serving. Transfer Engine is the core component of Mooncake, see [documentations](https://github.com/kvcache-ai/mooncake/blob/main/doc/en/transfer-engine.md) for its design & API list. 

Compared with NCCL, Mooncake Transfer Engine has the following features:
- a unified programming interface for data transfers between DRAM-to-DRAM (both local and remote), DRAM-to-GPU VRAM (both local and remote), and DRAM-to-remote NVMe devices
- support for TCP, RDMA, and NVMe-of protocols
- topology-aware path selection (link to our english doc, transfer_engine.md), aggregating bandwidth from multiple NICs

Like the current implementation of [PR #8498](https://github.com/vllm-project/vllm/pull/8498), there are two roles: KV provider (e.g. prefill vLLM instance) and KV consumer (e.g. decode vLLM instance)
- Provider side implements `insert`: insert a KV cache into a buffer, so that it can be transferred upon request
- Consumer side implements `drop_select`: select a KV cache based on tokens, transfer the selected KV, and drop this KV out from the buffer

Both roles are run in different machines.

Integration guide: https://github.com/kvcache-ai/mooncake/blob/main/doc/en/vllm-integration.md

Benchmark result: https://github.com/kvcache-ai/mooncake/blob/main/doc/en/vllm_benchmark_results.md

### CODE REVIEW

**Comment by stmatengss** [comment]
Currently, this PR is based on the early version of #8498. We plan to clean up and rebase the code against the latest version soon. Apologies for triggering the request review prematurely.

**Comment by KuntaiDu** [comment]
The new version of disaggregated prefill PR #10502  is just merged, and feel free to continue development in vLLM's main branch!  API-wise the new PR is pretty similar to the old PR so (hopefully) it is straightforward to migrate the implementation.

File: vllm/distributed/kv_transfer/kv_pipe/mooncake_distributed_pipe.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        """Load the config from a JSON file."""
+        with open(file_path, 'r') as fin:
+            config = json.load(fin)
+        return MooncakeTransferEngineConfig(
+            prefill_url=config.get("prefill_url"),
+            decode_url=config.get("decode_url"),
+            metadata_server=config.get("metadata_server"),
+            protocol=config.get("protocol", "tcp"),
+            device_name=config.get("device_name", ""),
+        )
```
Comment (junna2016) [inline vllm/distributed/kv_transfer/kv_pipe/mooncake_distributed_pipe.py:38]: Can you provide an test example to run disaggregated prefill/decoding mode with MooncakeDistributedPipe scene?

**Comment by junna2016** [comment]
Can you provide an test example to run disaggregated prefill/decoding mode with MooncakeDistributedPipe scene?

**Comment by ShangmingCai** [comment]
> Can you provide a test example to run disaggregated prefill/decoding mode with MooncakeDistributedPipe scene?

You can refer to [this doc](https://github.com/kvcache-ai/Mooncake/blob/main/doc/en/vllm-integration.md) to run a demo based on PR 8498. Currently, we are rebasing from the main branch. It is nearly done, but we will run more tests to ensure its compatibility.

**Comment by junna2016** [comment]
> > Can you provide a test example to run disaggregated prefill/decoding mode with MooncakeDistributedPipe scene?
> 
> You can refer to [this doc](https://github.com/kvcache-ai/Mooncake/blob/main/doc/en/vllm-integration.md) to run a demo based on PR 8498. Currently, we are rebasing from the main branch. It is nearly done, but we will run more tests to ensure its compatibility.

Thanks a lot

**Comment by ShangmingCai** [comment]
After rebase, we move the development to PR https://github.com/vllm-project/vllm/pull/10884 now.

**Comment by DarkLight1337** [comment]
Closing as superseded by #10884

**Pull request closed without merging**